hxwls.1
.TH "HXWLS" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"

.de d \" begin display
.sp
.in +4
.nf
.ft CR
.CDS
..
.de e \" end display
.CDE
.in -4
.fi
.ft R
.sp
..

.SH NAME
hxwls \- list links in an HTML file
.SH SYNOPSIS
.B hxwls
.RB "[\| " \-l " \|]"
.RB "[\| " \-t " \|]"
.RB "[\| " \-r " \|]"
.RB "[\| " \-h " \|]"
.RB "[\| " \-a " \|]"
.RB "[\| " \-b
.IR " base" " \|]"
.RI "[\| " file " \|]"
.SH DESCRIPTION
.LP
The
.B hxwls
command reads an HTML file (standard input by default) and prints out
all links it finds. The output is written to stdout.
.SH OPTIONS
The following options are supported:
.TP 10
.B \-l
Produce a long listing. Instead of just the URI,
.B hxwls
prints three columns: the element name, the value of the REL
attribute, and the target URI.
.TP
.B \-t
Produce a tuple listing.
.B hxwls
prints four columns: the URI of the document itself, the element name,
the value of the REL attribute, and the target URI.
.TP
.BI \-r
Print relative URLs as they are, without converting them to absolute
URLs.
.TP
.BI \-b " base"
Use
.I base
as the initial base URL. If there is a <base> element in the document, 
it will override the \-b option.
.TP
.B \-h
Output as HTML. The output will be listed in the form of <a> elements.
.TP
.B \-a
Convert any IRIs (Internationalized Resource Identifiers) to
ASCII-only URIs. This causes any non-ASCII characters in the path of a
URI to be encoded as %-escaped octets and non-ASCII characters in the
domain name as punycode. (Punycode encoding is only available if
.B hxwls
is compiled with libidn support.)
.SH OPERANDS
The following operand is supported:
.TP 10
.I file
The name or the URL of an HTML file. If absent, standard input is read instead.
.SH "DIAGNOSTICS"
The following exit values are returned:
.TP 10
.B 0
Successful completion.
.TP
.B > 0
An error occurred in the parsing of the HTML file.
.B hxwls
will try to correct the error and produce output anyway.
.SH "SEE ALSO"
.BR asc2xml (1),
.BR hxnormalize (1),
.BR hxnum (1),
.BR xml2asc (1)


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxwls.c
/*
 * List all links from the given document.
 *
 * Copyright © 1994-2000 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/copyright-software
 *
 * Bert Bos <bert@w3.org>
 * Created 31 July 1999
 * $Id: hxwls.c,v 1.13 2019/08/28 19:14:34 bbos Exp $
 */
#include "config.h"
#include <assert.h>
#ifdef HAVE_UNISTD_H
#  include <unistd.h>
#endif
#include <ctype.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#if STDC_HEADERS
# include <string.h>
#else
# ifndef HAVE_STRCHR
#  define strchr index
#  define strrchr rindex
# endif
# ifndef HAVE_STRDUP
#  include "strdup.e"
# endif
#endif
#include "export.h"
#include "heap.e"
#include "types.e"
#include "html.e"
#include "scan.e"
#include "dict.e"
#include "openurl.e"
#include "url.e"
#include "errexit.e"
#include "unent.e"

static bool has_error = false;
static string base = NULL;
static string self;
static enum {Short, Long, HTML, Tuple} format = Short;	/* Option -l -h -t */
static bool relative = false;		/* Option -r */
static bool ascii = false;		/* Option -a */


/* append_utf8 -- add UTF-8 bytes for code n at s, return end of string */
static string append_utf8(string s, const unsigned long n)
{
  /* We assume s is long enough */
  if (n <= 0x7F) {
    *(s++) = n;
  } else if (n <= 0x7FF) {
    *(s++) = 0xC0 | (n >> 6);
    *(s++) = 0x80 | (n & 0x3F);
  } else if (n <= 0xFFFF) {
    *(s++) = 0xE0 | (n >> 12);
    *(s++) = 0x80 | ((n >> 6) & 0x3F);
    *(s++) = 0x80 | (n & 0x3F);
  } else if (n <= 0x1FFFFF) {
    *(s++) = 0xF0 | (n >> 18);
    *(s++) = 0x80 | ((n >> 12) & 0x3F);
    *(s++) = 0x80 | ((n >> 6) & 0x3F);
    *(s++) = 0x80 | (n & 0x3F);
  } else if (n <= 0x3FFFFFF) {
    *(s++) = 0xF0 | (n >> 24);
    *(s++) = 0x80 | ((n >> 18) & 0x3F);
    *(s++) = 0x80 | ((n >> 12) & 0x3F);
    *(s++) = 0x80 | ((n >> 6) & 0x3F);
    *(s++) = 0x80 | (n & 0x3F);
  } else {
    *(s++) = 0xF0 | (n >> 30);
    *(s++) = 0x80 | ((n >> 24) & 0x3F);
    *(s++) = 0x80 | ((n >> 18) & 0x3F);
    *(s++) = 0x80 | ((n >> 12) & 0x3F);
    *(s++) = 0x80 | ((n >> 6) & 0x3F);
    *(s++) = 0x80 | (n & 0x3F);
  }
  return s;
}


/* output -- print the link (lowercases rel argument) */
static void output(const conststring type, const conststring rel,
		   conststring url)
{
  string h = NULL, q, r, rel1;
  conststring p, s;
  const struct _Entity *e;
  unsigned long c;

  if (url) {					/* If we found a URL */

    /* Replace entities. */
    h = newnstring(url, 2 * strlen(url));	/* Reserve sufficient space */
    for (p = url, q = h; *p; p++) {
      if (*p != '&') {
	*(q++) = *p;
      } else if (*(p+1) == '#') {		/* Numeric entity */
	if (*(p+2) == 'x') c = strtoul(p + 3, &r, 16);
	else c = strtoul(p + 2, &r, 10);
	if (c > 0 && c <= 2147483647) q = append_utf8(q, c);
	p = *r == ';' ? r : r - 1;
      } else {					/* Entity */
	for (s = p + 1; isalnum(*s); s++);
	if (!(e = lookup_entity(p+1, s - (p+1)))) *(q++) = '&'; /* Unknown */
	else {q = append_utf8(q, e->code); p = *s == ';' ? s : s -1;}
      }
    }
    *q = '\0';
    url = h;

    /* Make URL absolute */
    if (! relative && base) {
      h = URL_s_absolutize(base, url);
      dispose(url);
      url = h;
    }
    /* Convert IRI to URL, if requested */
    if (ascii) {
      h = URL_s_to_ascii(url);
      dispose(url);
      url = h;
    }
    rel1 = newstring(rel ? rel : "");
    down(rel1);
    switch (format) {
      case HTML:
	printf("<li><a class=\"%s\" rel=\"%s\" href=\"%s\">%s</a></li>\n",
		 type, rel1, url, url);
	break;
      case Long:
	printf("%s\t%s\t%s\n", type, rel1, url);
	break;
      case Short:
	printf("%s\n", url);
	break;
      case Tuple:
	printf("%s\t%s\t%s\t%s\n", self, type, rel1, url);
	break;
      default:
	assert(!"Cannot happen!");
    }
    free(rel1);
    free(h);
  }
}


/* --------------- implements parser interface api------------------------- */

/* handle_error -- called when a parse error occurred */
void handle_error(void *clientdata, const string s, int lineno)
{
  fprintf(stderr, "%d: %s\n", lineno, s);
  has_error = true;
}

/* start -- called before the first event is reported */
void* start(void)
{
  if (format == HTML) {
    printf("<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.0//EN\"\n");
    printf("  \"http://www.w3.org/TR/REC-html40/strict.dtd\">\n");
    printf("<html>\n");
    printf("<head><title>Output of listlinks</title></head>\n");
    printf("<body>\n");
    printf("<ol>\n");
  }
  return NULL;
}

/* end -- called after the last event is reported */
void end(void *clientdata)
{
  if (format == HTML) {
    printf("</ol>\n");
    printf("</body>\n");
    printf("</html>\n");
  }
}

/* handle_comment -- called after a comment is parsed */
void handle_comment(void *clientdata, string commenttext)
{
  free(commenttext);
}

/* handle_text -- called after a text chunk is parsed */
void handle_text(void *clientdata, string text)
{
  /* There may be several consecutive calls to this routine. */
  /* escape(text); */
  free(text);
}

/* handle_decl -- called after a declaration is parsed */
void handle_decl(void *clientdata, string gi, string fpi, string url)
{
  /* skip */
  if (gi) free(gi);
  if (fpi) free(fpi);
  if (url) free(url);
}

/* handle_pi -- called after a PI is parsed */
void handle_pi(void *clientdata, string pi_text)
{
  if (pi_text) free(pi_text);
}

/* handle_starttag -- called after a start tag is parsed */
void handle_starttag(void *clientdata, string name, pairlist attribs)
{
  /* ToDo: print text of anchor, if available */
  conststring h;

  if (strcasecmp(name, "base") == 0) {
    h = pairlist_get(attribs, "href");
    if (h) base = strdup(h);			/* Use as base from now on */
    output("base", NULL, h);
  } else if (strcasecmp(name, "link") == 0) {
    output("link", pairlist_get(attribs, "rel"), pairlist_get(attribs, "href"));
  } else if (strcasecmp(name, "a") == 0) {
    output("a", pairlist_get(attribs, "rel"), pairlist_get(attribs, "href"));
  } else if (strcasecmp(name, "img") == 0) {
    output("img", NULL, pairlist_get(attribs, "src"));
    output("img", "longdesc", pairlist_get(attribs, "longdesc"));
    output("img", "srcset", pairlist_get(attribs, "srcset"));
  } else if (strcasecmp(name, "input") == 0) {
    output("input", "src", pairlist_get(attribs, "src"));
  } else if (strcasecmp(name, "object") == 0) {
    output("object", NULL,  pairlist_get(attribs, "data"));
    output("object", "classid",  pairlist_get(attribs, "classid"));
    output("object", "codebase",  pairlist_get(attribs, "codebase"));
  } else if (strcasecmp(name, "area") == 0) {
    output("area", pairlist_get(attribs, "rel"), pairlist_get(attribs, "href"));
  } else if (strcasecmp(name, "ins") == 0) {
    output("ins", NULL, pairlist_get(attribs, "cite"));
  } else if (strcasecmp(name, "del") == 0) {
    output("del", NULL, pairlist_get(attribs, "cite"));
  } else if (strcasecmp(name, "q") == 0) {
    output("q", NULL, pairlist_get(attribs, "cite"));
  } else if (strcasecmp(name, "blockquote") == 0) {
    output("bq", NULL, pairlist_get(attribs, "cite"));
  } else if (strcasecmp(name, "form") == 0) {
    output("form", pairlist_get(attribs, "method"), pairlist_get(attribs, "action"));
  } else if (strcasecmp(name, "frame") == 0) {
    output("frame", NULL, pairlist_get(attribs, "src"));
  } else if (strcasecmp(name, "iframe") == 0) {
    output("iframe", NULL, pairlist_get(attribs, "src"));
  } else if (strcasecmp(name, "head") == 0) {
    output("head", NULL, pairlist_get(attribs, "profile"));
  } else if (strcasecmp(name, "script") == 0) {
    output("script", NULL, pairlist_get(attribs, "src"));
  } else if (strcasecmp(name, "body") == 0) {
    output("body", NULL, pairlist_get(attribs, "background"));
  } else if (strcasecmp(name, "video") == 0) {
    output("video", NULL, pairlist_get(attribs, "src"));
  } else if (strcasecmp(name, "audio") == 0) {
    output("audio", NULL, pairlist_get(attribs, "src"));
  } else if (strcasecmp(name, "source") == 0) {
    output("source", "srcset", pairlist_get(attribs, "srcset"));
    output("source", "src", pairlist_get(attribs, "src"));
  }

  /* Free memory */
  pairlist_delete(attribs);
  free(name);
}

/* handle_emptytag -- called after an empty tag is parsed */
void handle_emptytag(void *clientdata, string name, pairlist attribs)
{
  handle_starttag(clientdata, name, attribs);
}

/* handle_endtag -- called after an endtag is parsed (name may be "") */
void handle_endtag(void *clientdata, string name)
{
  free(name);
}

/* --------------------------------------------------------------------- */


/* usage -- print usage message and exit */
static void usage(string progname)
{
  fprintf(stderr,
	  "Version %s\nUsage: %s [-l] [-r] [-h] [-b base] [-t] [-a] [HTML-file]\n",
	  VERSION, progname);
  exit(1);
}


int main(int argc, char *argv[])
{
  int c, status = 200;

  /* Bind the parser callback routines to our handlers */
  set_error_handler(handle_error);
  set_start_handler(start);
  set_end_handler(end);
  set_comment_handler(handle_comment);
  set_text_handler(handle_text);
  set_decl_handler(handle_decl);
  set_pi_handler(handle_pi);
  set_starttag_handler(handle_starttag);
  set_emptytag_handler(handle_emptytag);
  set_endtag_handler(handle_endtag);

  /* Parse command line arguments */
  while ((c = getopt(argc, argv, "lb:rhta")) != -1) {
    switch (c) {
      case 'l': format = Long; break;		/* Long listing */
      case 'b': base = strdup(optarg); break;	/* Set base of URL */
      case 'r': relative = true; break;		/* Do not make URLs absolute */
      case 'h': format = HTML; break;		/* Output in HTML format */
      case 't': format = Tuple; break;		/* Output as 4-tuples */
      case 'a': ascii = true; break;		/* Convert IRIs to URLs */
      default: usage(argv[0]);
    }
  }

  if (optind == argc) {
    yyin = stdin;
    self = "-";
  } else if (optind == argc - 1) {
    if (!base) base = strdup(argv[optind]);
    if (eq(argv[optind], "-")) yyin = stdin;
    else yyin = fopenurl(argv[optind], "r", &status);
    self = argv[optind];
  } else {
    usage(argv[0]);
  }

  if (yyin == NULL) {perror(argv[optind]); exit(1);}
  if (status != 200) errexit("%s : %s\n", argv[optind], http_strerror(status));

  if (yyparse() != 0) exit(3);

  if (base) free(base);

  return has_error ? 1 : 0;
}
